File: sdks/java/src/main/java/com/blok/nanoservice/logging/LogEntry.java
package com.blok.nanoservice.logging;

import com.google.gson.Gson;

import java.time.Instant;
import java.util.Map;
import java.util.Objects;

/**
 * Represents a single structured log entry.
 */
public class LogEntry {

    private final LogLevel level;
    private final String message;
    private final Instant timestamp;
    private final Map<String, Object> fields;

    public LogEntry(LogLevel level, String message, Instant timestamp, Map<String, Object> fields) {
        this.level = level != null ? level : LogLevel.INFO;
        this.message = message != null ? message : "";
        this.timestamp = timestamp != null ? timestamp : Instant.now();
        this.fields = fields;
    }

    public LogLevel getLevel() {
        return level;
    }

    public String getMessage() {
        return message;
    }

    public Instant getTimestamp() {
        return timestamp;
    }

    public Map<String, Object> getFields() {
        return fields;
    }

    /**
     * Formats this entry as a human-readable string.
     *
     * @return formatted log line
     */
    public String format() {
        StringBuilder sb = new StringBuilder();
        sb.append("[").append(level).append("] ");
        sb.append(timestamp.toString()).append(" ");
        sb.append(message);
        if (fields != null && !fields.isEmpty()) {
            sb.append(" ").append(new Gson().toJson(fields));
        }
        return sb.toString();
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        LogEntry logEntry = (LogEntry) o;
        return level == logEntry.level &&
                Objects.equals(message, logEntry.message) &&
                Objects.equals(timestamp, logEntry.timestamp) &&
                Objects.equals(fields, logEntry.fields);
    }

    @Override
    public int hashCode() {
        return Objects.hash(level, message, timestamp, fields);
    }

    @Override
    public String toString() {
        return format();
    }
}

File: sdks/php/src/Logging/LogEntry.php
<?php

declare(strict_types=1);

namespace Blok\Nanoservice\Logging;

/**
 * LogEntry represents a single log entry.
 */
final class LogEntry
{
    public function __construct(
        public readonly LogLevel $level,
        public readonly string $message,
        public readonly string $timestamp,
        public readonly ?array $fields = null,
    ) {}

    /**
     * Format the entry as a human-readable string.
     */
    public function __toString(): string
    {
        $base = sprintf('[%s] %s %s', $this->level->value, $this->timestamp, $this->message);
        if ($this->fields !== null) {
            $json = json_encode($this->fields, JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_UNICODE);
            return $base . ' ' . ($json ?: '{}');
        }
        return $base;
    }

    /**
     * Convert to an associative array for JSON serialization.
     */
    public function toArray(): array
    {
        $result = [
            'level' => $this->level->value,
            'message' => $this->message,
            'timestamp' => $this->timestamp,
        ];
        if ($this->fields !== null) {
            $result['fields'] = $this->fields;
        }
        return $result;
    }
}
